app/components/new-user.js

```diff
@@ -1,78 +1,166 @@
-    validator('presence', true),
-    validator('length', {
-      max: 100
-    }),
-    validator('format', {
-      type: 'email'
-    })
-  ],
-  phone: [
-    validator('length', {
-      max: 20
-    })
-  ]
-});
-
-export default Component.extend(NewUser, Validations, {
-  classNames: ['new-user'],
-
-  keyUp(event) {
-    const keyCode = event.keyCode;
-    const target = event.target;
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { inject as service } from '@ember/service';
+import {all, filter} from 'rsvp';
+import {dropTask, restartableTask} from 'ember-concurrency-decorators';
+import moment from "moment";
+import { validatable, IsEmail, Length, NotBlank } from 'ilios-common/decorators/validation';
 
-    if ('text' !== target.type) {
-      return;
+@validatable
+export default class NewUserComponent extends Component {
+  @service store;
+  @service currentUser;
+  @service flashMessages;
+  @service permissionChecker;
+
+  @tracked @Length(1, 50) @NotBlank() firstName = null;
+  @tracked @Length(1, 20) middleName = null;
+  @tracked @Length(1, 50) @NotBlank() lastName = null;
+  @tracked @Length(1, 16) campusId = null;
+  @tracked @Length(1, 16) otherId = null;
+  @tracked @IsEmail() @Length(1, 100) @NotBlank() email = null;
+  @tracked @Length(1, 100) @NotBlank() username = null;
+  @tracked @NotBlank() password = null;
+  @tracked @Length(1, 20) phone = null;
+  @tracked schoolId = null;
+  @tracked primaryCohortId = null;
+  @tracked primarySchool = null;
+  @tracked cohorts = [];
+  @tracked schools = [];
+  @tracked isSaving = false;
+  @tracked nonStudentMode = true;
+
+  get bestSelectedSchool() {
+    if (this.schoolId) {
+      const currentSchool = this.schools.findBy('id', this.schoolId);
+
+      if (currentSchool) {
+        return currentSchool;
+      }
     }
+    return this.primarySchool;
+  }
 
-    if (13 === keyCode) {
-      this.save.perform();
+  get bestSelectedCohort() {
+    if (! this.currentSchoolCohorts) {
+      return null;
+    }
+
+    if (this.primaryCohortId) {
+      const currentCohort = this.currentSchoolCohorts.findBy('id', this.primaryCohortId);
+
+      if (currentCohort) {
+        return currentCohort;
+      }
+    }
+
+    return this.currentSchoolCohorts.lastObject;
+  }
+
+  @restartableTask
+  *load(){
+    const user = yield this.currentUser.model;
+    this.primarySchool = yield user.school;
+    this.schools = yield this.loadSchools.perform();
+    this.currentSchoolCohorts = yield this.bestSelectedSchool?.cohorts;
+    this.cohorts = yield this.loadCohorts.perform(this.bestSelectedSchool);
+  }
+
+  @restartableTask
+  *reload() {
+    this.currentSchoolCohorts = yield this.bestSelectedSchool?.cohorts;
+    this.cohorts = yield this.loadCohorts.perform(this.bestSelectedSchool);
+  }
+
+  @restartableTask
+  *loadSchools() {
+    const store = this.store;
+    const schools = yield store.findAll('school', { reload: true });
```

Comment:
An improvement may be possible here, because we always load all schools in https://github.com/ilios/frontend/blob/master/app/routes/application.js#L45 `peekRecord` could be used here synchronously replacing this with a simple `get schools` and possible making all async in this component a bit easier.

Reply:
i changed it. the method will still have to be declared as either a task or an async function, since the perms check that follow the fetching of schools are async in nature.

Reply:
Oh, yea. I see that now. Ok. Accepted.